### ISSUE TITLE
ZFS kernel module missing from 5.15

### ISSUE THREAD

**Issue by DCrompton324** [pop-os/linux#88]
**How did you upgrade to 21.10? (Fresh install / Upgrade)**

 Fresh install, all packages up to date at time of post
 
**Related Application and/or Package Version (run `apt policy $PACKAGE NAME`):**

zfsutils-linux:
  Installed: 2.0.6-1ubuntu2
  Candidate: 2.0.6-1ubuntu2

**Issue/Bug Description:**
When installing zfs utils the following error is received

modprobe: FATAL: Module zfs not found in directory /lib/modules/5.15.0-76051500-generic
zfs-import-scan.service is a disabled or a static unit, not starting it.

Following this zfs volumes seem mounted but their contents appear missing

**Steps to reproduce (if you know):**
sudo apt install zfsutils-linux

**Expected behavior:**
Kernel module loads and I can access zfs volumes

**Other Notes:**
Works when I load the old 5.13.0-7260 kernel, I can then mount and access zfs volumes. 

zfs module is present at /lib/modules/5.13.0-7620-generic/kernel/zfs

There is no /lib/modules/5.15.0-76051500-generic/kernel/zfs

**Comment by jkhii** [comment]
This also breaks on 20.04 after recent update. I had to boot to 5.13 kernel.

**Comment by jgust** [comment]
> I'm on 21.04 but the solution given by jgust does not work for me: the version of zfsutils-linux in 21.04 conflicts with the version required by zfs-dkms in 21.10. As of now my system is broken and I'm not able to mount my home directory.

@anahata0108,  Weird that you had a different behavior on 21.04.  After installing the packages i mentioned previously I end up with:
```
dpkg -s zfsutils-linux
Package: zfsutils-linux
Status: install ok installed
Priority: optional
Section: contrib/admin
Installed-Size: 1511
Maintainer: Ubuntu Developers <ubuntu-devel-discuss@lists.ubuntu.com>
Architecture: amd64
Source: zfs-linux
Version: 2.0.2-1ubuntu5.3
Replaces: spl (<< 0.7.9-2), spl-dkms
Provides: zfsutils
Depends: libnvpair3linux (= 2.0.2-1ubuntu5.3), libuutil3linux (= 2.0.2-1ubuntu5.3), libzfs4linux (= 2.0.2-1ubuntu5.3), libzpool4linux (= 2.0.2-1ubuntu5.3), python3, libblkid1 (>= 2.16), libc6 (>= 2.33), libuuid1 (>= 2.16)
Pre-Depends: init-system-helpers (>= 1.54~)
Recommends: lsb-base, zfs-zed
Suggests: nfs-kernel-server, samba-common-bin (>= 3.0.23), zfs-initramfs | zfs-dracut
Breaks: openrc, spl (<< 0.7.9-2), spl-dkms (<< 0.8.0~rc1), zfs-dkms (>> 2.0.2-1ubuntu5.3...), zfs-dkms (<< 2.0.2-1ubuntu5.3)
Conflicts: zfs, zfs-fuse
Conffiles:
 /etc/cron.d/zfsutils-linux 988b0007fc4d08bb239f7857d7403c87
 /etc/default/zfs 75603cca7a73a251e272024c5f06b93e
 /etc/init.d/zfs-import 4ed1fff952ac0641ff6650dad1983623
 /etc/init.d/zfs-mount 5c235ec583e083d0eac63c5060856c42
 /etc/init.d/zfs-share 40cf9308ba8681eea005fbd06bce3785
 /etc/sudoers.d/zfs ad829cd4055ab2a8d41f8b5f2b564a3f
 /etc/zfs/zfs-functions a02507b820abff58a2f9f0b758a3040b
Description: command-line tools to manage OpenZFS filesystems
 OpenZFS is a storage platform that encompasses the functionality of
 traditional filesystems and volume managers. It supports data checksums,
 compression, encryption, snapshots, and more.
 .
 This package provides the zfs and zpool commands to create and administer
 OpenZFS filesystems.
Homepage: https://zfsonlinux.org/
Original-Maintainer: Debian ZFS on Linux maintainers <pkg-zfsonlinux-devel@alioth-lists.debian.net>
```
``` 
dpkg -s zfs-dkms
Package: zfs-dkms
Status: deinstall ok config-files
Priority: optional
Section: contrib/kernel
Installed-Size: 17127
Maintainer: Ubuntu Developers <ubuntu-devel-discuss@lists.ubuntu.com>
Architecture: all
Source: zfs-linux
Version: 2.0.6-1ubuntu2.1
Config-Version: 2.0.6-1ubuntu2.1
Replaces: spl-dkms
Provides: zfs-modules
Depends: dkms (>> 2.1.1.2-5), file, libc6-dev | libc-dev, lsb-release, python3-distutils | libpython3-stdlib (<< 3.6.4), debconf (>= 0.5) | debconf-2.0, perl:any
Recommends: zfs-zed, zfsutils-linux (>= 2.0.6-1ubuntu2.1), linux-libc-dev (<< 5.15~), linux-libc-dev (>= 3.10~)
Suggests: debhelper
Breaks: spl-dkms (<< 0.8.0~rc1)
Description: OpenZFS filesystem kernel modules for Linux
 OpenZFS is a storage platform that encompasses the functionality of
 traditional filesystems and volume managers. It supports data checksums,
 compression, encryption, snapshots, and more.
 .
 This DKMS package includes the SPA, DMU, ZVOL, and ZPL components of
 OpenZFS.
Homepage: https://zfsonlinux.org/
Original-Maintainer: Debian ZFS on Linux maintainers <pkg-zfsonlinux-devel@alioth-lists.debian.net>
```
As you can see, neither depends on the other so its strange you were getting conflicts. Are you on the same versions?

One step I realize I forgot to mention was after removing `zfs-dkms`and `zfsutils-linux` I also did a `sudo apt autoremove` which removed other zfs releated packages that were no longer in use.  These were then automatically re-installed when re-installing `zfsutils-linux`

**Comment by anahata0108** [comment]
Regardless of which versions I have, installing a package from 21.10 shouldn't be necessary to upgrade on 21.04 or 20.04. This is clearly a kluge workaround and not the correct way to address the problem.

**Comment by jgust** [comment]
You are correct, this is by no means an official (or proper) solution. I'm sorry if you somehow got that impression. I'm sure the correct solution is being worked on as we speak and will be available in due time. In case anyone wants to get this working on 21.04 in the mean time, this is a solution that worked great for me, YMMV.

**Comment by LowryConor** [comment]
FWIW the workaround above did not work for me on 21.04 either. Once I re-install zfs-utils, the 2.0.6 version of zfs-dkms is removed and gets replaced by 2.0.2

```
sudo apt install zfsutils-linux --no-remove
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following additional packages will be installed:
  libnvpair3linux libuutil3linux libzfs4linux libzpool4linux zfs-zed
Suggested packages:
  nfs-kernel-server samba-common-bin zfs-initramfs | zfs-dracut
The following packages will be REMOVED:
  zfs-dkms
The following NEW packages will be installed:
  libnvpair3linux libuutil3linux libzfs4linux libzpool4linux zfs-zed zfsutils-linux
0 upgraded, 6 newly installed, 1 to remove and 1 not upgraded.
E: Packages need to be removed but remove is disabled.
```

**Comment by anahata0108** [comment]
This is what I see as well, but I wasn't interested in breaking my system again to copy-paste the error messages.

**Comment by jgust** [comment]
Fair enough, maybe I got lucky :-) Perhaps a bit more robust solution that I found is to add a ppa that provides the latest zfs packages:
```
sudo add-apt-repository ppa:jonathonf/zfs
sudo apt update
sudo apt install zfsutils-linux zfs-dkms
```
[source](https://blog.uptrace.dev/posts/ubuntu-install-zfs.html)

Of course, if you don't mind staying on 5.13 you probably shouldn't bother. But for me 5.15 supports my hardware much better than 5.13 and earlier so I'm really keen using it.

**Comment by cinimodev** [comment]
Been following this thread for a couple days and having the same issue as @LowryConor. Both `zfs-dkms` and `zfsutils-linux` install version 2.02, which is only supported to 5.14. I'm hesitant to install any PPA's, etc. for a kludge fix until the next update comes. In exchange, my data drives now don't mount because zfs is broken until either an update comes for 5.15 or I decide to just run an older kernel.

**Comment by anahata0108** [comment]
Is there anyone actually looking at this? People's systems are broken and we are resorting to ugly hacks to fix this, yet there's been no acknowledgement from system76 that this is even an issue.

**Comment by cinimodev** [comment]
It looks like OpenZFS version 2.0.7 is ready for kernel 5.15. But, I can't have a distro that will break my zfs pools. I agree with @anahata0108  that I have also been hesitant to update my system now due to specific settings I need to get my work done. I have been putting it off, but it seems like Pop isn't the distro for me anymore and will have to go back to something that wants to support zfs as a file system option. Sad day, but think its my only option now.

**Comment by cinimodev** [comment]
I want to follow up on this. It appears `zfs-dkms` will install correctly after upgrading to Pop 21.10. I was previously on 21.04 and before wiping for a different distro I upgraded, then re-installed both `zfsutils-linux` and `zfs-dkms`. After installing, `modprobe zfs` and then re-import the pools. After doing this, everything is working. 

Nothing I did on 21.04 would work. However, after upgrading to 21.10 OpenZFS installed correctly. 

I am still concerned that with frequent kernel updates that this will be a recurring issue. Personally, I would prefer to stay on the LTS kernel for my primary workstation to avoid update breakage.

**Comment by rjpower4** [comment]
Continuing with what @cinimodev mentioned, running zfs 2.0.6 with kernel 5.15 is **unsupported**, it would be great to get access to zfs 2.0.7 on 21.10

**Comment by jacobgkau** [comment]
> Nothing I did on 21.04 would work. However, after upgrading to 21.10 OpenZFS installed correctly.

This is because the version of `zfs-dkms` being shipped by Ubuntu on 21.04 was too old for kernel 5.15, but the version being shipped on version 21.10 was new enough to work. We have backported the 21.10 version to 20.04 LTS for Pop!_OS (since we ship the same kernel on both releases), so both are currently working.

> I would prefer to stay on the LTS kernel for my primary workstation to avoid update breakage.

5.15 is listed by the upstream kernel developers as the latest LTS kernel: https://www.kernel.org/category/releases.html (although we want to ship the latest mainline kernel possible anyway.)

I am closing this issue because we are now [packaging ZFS](https://github.com/pop-os/zfs-linux/) and regression testing that it works for every kernel update. This should address any concerns that ZFS will break with Pop!_OS's more frequent kernel updates.

> Continuing with what @cinimodev mentioned, running zfs 2.0.6 with kernel 5.15 is unsupported, it would be great to get access to zfs 2.0.7 on 21.10

If you know of any specific functionality that is not working, please let us know. We will add your case to our regression testing procedures and re-open this issue.
